File: [0046][Medium][Permutations][backtracking]/Permutations.py
# leetcode time     cost : 44 ms
# leetcode memory   cost : 13.9 MB
# Time  Complexity: O(n*n!)
# Space Complexity: O(n)
# solution 1,backtracking
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack(first = 0):
            # 所有数都填完了
            if first == n:  
                res.append(nums[:])
            for i in range(first, n):
                # 动态维护数组
                nums[first], nums[i] = nums[i], nums[first]
                # 继续递归填下一个数
                backtrack(first + 1)
                # 撤销操作
                nums[first], nums[i] = nums[i], nums[first]
        
        n = len(nums)
        res = []
        backtrack()
        return res
def main():
    inputX,expectRes = [1,2,3], [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,2,1],[3,1,2]]
    obj = Solution()
    result = obj.permute(inputX)
    try:
        assert result == expectRes
        print("passed, result is follow expect:",result)
    except AssertionError as aError:
        print('failed, result >> ', result,"<< is wrong, ","expect is : "+ expectRes)
    
if __name__ =='__main__':
    main()  

File: [0046][Medium][Permutations][backtracking]/Permutations.cpp
// -*- coding: utf-8 -*- 
// leetcode time     cost : 12 ms
// leetcode memory   cost : 8 MB
// Time  Complexity: O(N*N!)
// Space Complexity: O(N*N!)
//https://leetcode-cn.com/problems/permutations/solution/hui-su-suan-fa-python-dai-ma-java-dai-ma-by-liweiw/
// solution 1,backtrack with used notation
#include <iostream>
#include <vector>
#include <set>
#include <stdlib.h>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        permuteCore(nums);
        return results;
    }

    void permuteCore(vector<int>& nums) {
        if (used.size() == nums.size()) {  // 所有元素已被访问过
            results.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (used.find(nums[i]) == used.end()) {  // 当前元素未被访问时
                path.push_back(nums[i]);
                used.insert(nums[i]);
                permuteCore(nums);
                used.erase(nums[i]);
                path.pop_back();
            }
        }
    }

private:
    vector<int> path;  // 记录当前的排列
    set<int> used;  // 标记已在排列中的元素
    vector<vector<int>> results;  // 最终的结果
};